src/client/pmix_client_get.c

```diff
@@ -497,64 +335,192 @@ static void getnb_cbfunc(struct pmix_peer_t *pr, pmix_usock_hdr_t *hdr,
     }
 }
 
-static void getnb_shortcut(int fd, short flags, void *cbdata)
+static void _getnbfn(int fd, short flags, void *cbdata)
 {
     pmix_cb_t *cb = (pmix_cb_t*)cbdata;
-    pmix_value_t val;
+    pmix_cb_t *cbret;
+    pmix_buffer_t *msg;
+    pmix_value_t *val;
     pmix_status_t rc;
-    int32_t m;
+    char *nm;
+    pmix_nspace_t *ns, *nptr;
+    size_t n;
 
-    pmix_output_verbose(2, pmix_globals.debug_output,
-                        "getnb_shortcut called with %s cbfunc",
-                        (NULL == cb->value_cbfunc) ? "NULL" : "NON-NULL");
+    /* if the nspace is empty, then the caller is referencing
+     * our own nspace */
+    if (0 == strlen(cb->procs->nspace)) {
+        nm = pmix_globals.myid.nspace;
+    } else {
+        nm = (char*)cb->procs->nspace;
+    }
 
-    PMIX_VALUE_CONSTRUCT(&val);
-    if (NULL != cb->value_cbfunc) {
-        m=1;
-        rc = pmix_bfrop.unpack(&cb->data, &val, &m, PMIX_VALUE);
-        cb->value_cbfunc(rc, &val, cb->cbdata);
+    /* find the nspace object */
+    nptr = NULL;
+    PMIX_LIST_FOREACH(ns, &pmix_globals.nspaces, pmix_nspace_t) {
+        if (0 == strcmp(nm, ns->nspace)) {
+            nptr = ns;
+            break;
+        }
+    }
+    if (NULL == nptr) {
+        /* we are asking for info about a new nspace - give us
+         * a chance to learn about it from the server. If the
+         * server has never heard of it, the server will return
+         * an error */
+         nptr = PMIX_NEW(pmix_nspace_t);
+         (void)strncpy(nptr->nspace, nm, PMIX_MAX_NSLEN);
+         pmix_list_append(&pmix_globals.nspaces, &nptr->super);
+         /* there is no point in looking for data in this nspace
+          * object, so let's just go generate the request */
+         goto request;
     }
-    PMIX_VALUE_DESTRUCT(&val);
-    PMIX_RELEASE(cb);
-}
 
-static void getnb_request(int fd, short flags, void *cbdata)
-{
-    get_request_caddy_t *data = (get_request_caddy_t *)cbdata;
-    pmix_cb_t *cb;
-    pmix_buffer_t *msg;
+    /* the requested data could be in the job-data table, so let's
+     * just check there first.  */
+    if (PMIX_SUCCESS == (rc = pmix_hash_fetch(&nptr->internal, PMIX_RANK_WILDCARD, cb->key, &val))) {
+        /* found it - we are in an event, so we can
+         * just execute the callback */
+        cb->value_cbfunc(rc, val, cb->cbdata);
+        /* cleanup */
+        if (NULL != val) {
+            PMIX_VALUE_RELEASE(val);
+        }
+        /* protect the data */
+        cb->procs = NULL;
+        cb->key = NULL;
+        cb->info = NULL;
+        PMIX_RELEASE(cb);
+        return;
+    }
+    if (PMIX_RANK_WILDCARD == cb->procs->rank) {
```

Comment:
@rhc54 here cb->proc->rank is used (and we don't actually know if proc still exists).